File: src-tauri/src/ws_server/dispatch.rs
//! Command dispatch for the WebSocket server.
//!
//! Routes incoming WS commands to domain-specific dispatch modules.
//! Each domain (fs, git, worktree, misc) lives in its own file to keep
//! dispatch logic under 250 lines per file.

use super::types::WsResponse;
use super::WsState;

/// Dispatch a command by name, deserializing args from JSON.
///
/// Returns a WsResponse with either the result or an error message.
pub async fn dispatch(
    id: u64,
    cmd: &str,
    args: serde_json::Value,
    state: &WsState,
) -> WsResponse {
    match dispatch_inner(cmd, args, state).await {
        Ok(value) => WsResponse::ok(id, value),
        Err(err) => WsResponse::err(id, err),
    }
}

/// Inner dispatch that routes to domain-specific handlers by command prefix.
async fn dispatch_inner(
    cmd: &str,
    args: serde_json::Value,
    state: &WsState,
) -> Result<serde_json::Value, String> {
    // Route by prefix to keep the top-level match small
    if cmd.starts_with("agent_") {
        return super::dispatch_agent::dispatch(cmd, args, state).await;
    }
    if cmd.starts_with("fs_") {
        return super::dispatch_fs::dispatch(cmd, args, state).await;
    }
    if cmd.starts_with("git_") {
        return super::dispatch_git::dispatch(cmd, args).await;
    }
    if cmd.starts_with("worktree_") {
        return super::dispatch_worktree::dispatch(cmd, args).await;
    }

    // Everything else goes through the misc dispatcher
    super::dispatch_misc::dispatch(cmd, args, state).await
}

File: sidecar/src/dispatch.ts
/**
 * Command dispatch router for the WebSocket server.
 *
 * Routes incoming WS commands to domain-specific dispatch modules by prefix.
 */

import type { SidecarState } from "./state.js";
import { dispatchFs } from "./dispatch/dispatch-fs.js";
import { dispatchGit } from "./dispatch/dispatch-git.js";
import { dispatchWorktree } from "./dispatch/dispatch-worktree.js";
import { dispatchAgent } from "./dispatch/dispatch-agent.js";
import { dispatchMisc } from "./dispatch/dispatch-misc.js";
import { dispatchTerminal } from "./dispatch/dispatch-terminal.js";

/** Terminal and file watcher command names (no shared prefix). */
const TERMINAL_COMMANDS = new Set([
  "spawn_terminal",
  "write_terminal",
  "resize_terminal",
  "kill_terminal",
  "kill_terminals_by_cwd",
  "list_terminals",
  "start_watch",
  "stop_watch",
  "list_watches",
]);

/**
 * Dispatch a command by name, routing to domain-specific handlers.
 *
 * Returns the result value on success, throws on error.
 */
export async function dispatch(
  cmd: string,
  args: Record<string, unknown>,
  state: SidecarState,
): Promise<unknown> {
  if (cmd.startsWith("agent_")) {
    return dispatchAgent(cmd, args, state);
  }
  if (cmd.startsWith("fs_")) {
    return dispatchFs(cmd, args, state);
  }
  if (cmd.startsWith("git_")) {
    return dispatchGit(cmd, args);
  }
  if (cmd.startsWith("worktree_")) {
    return dispatchWorktree(cmd, args);
  }
  if (TERMINAL_COMMANDS.has(cmd)) {
    return dispatchTerminal(cmd, args, state);
  }
  return dispatchMisc(cmd, args, state);
}
